[1.5.0-rc] feat: 增加文档阅读器

JinliG opened the pull request:
## 主要改动
### 新增文档阅读模式
- 支持上传文档类进行侧边阅读
- 可以在对话时随时取消文档上下文关联
- 可以在阅读区手动选择页码，将指定内容关联到对话上下文

## 期望解决什么问题
- 在使用模型帮助阅读一些文档时，由于文档大小不可控，默认的 `file.read` 可能会导致 token 超出模型服务上线。
- 在使用模型辅助阅读专业文档（例如 财报、论文时），很可能并不需要将全部上下文提供给模型，用户可以自行决定把哪些内容提供给模型，来提供帮助。

## 演示
在助手设置中，新增【阅读器设置】，上传 或者 从文件中选择要阅读的文档（目前仅支持 pdf）。
![image](https://github.com/user-attachments/assets/0bd43c1d-3128-4133-8d99-f5b8c5b5a31e)


将在对话区左侧新增一个阅读区展示文档内容，并允许用户开关是否在对话中关联该文档，或选择哪几页进行关联。
![image](https://github.com/user-attachments/assets/e6406cb3-c831-4718-999a-11cf5b14c58e)






## 具体改动

**助手设置**
- 新增 AssistantDocumentSettings 阅读器设置

**对话框**
- 新增 PdfReader，Chat 注册宽度计算事件
- 调整 AttachmentPreview，增加支持的预览元素

JinliG committed:
feat: add document reader

kangfenmao commented:
文档预览放在右侧是不是比较好

JinliG commented:
> 猫哥给的方案 :-) <img alt="cb1dec80d0d68fa2129b98bda79995df" width="600" src="https://private-user-images.githubusercontent.com/144885467/436605814-8b938605-754f-49e9-baff-7c47909e268c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0NzU4ODksIm5iZiI6MTc0NTQ3NTU4OSwicGF0aCI6Ii8xNDQ4ODU0NjcvNDM2NjA1ODE0LThiOTM4NjA1LTc1NGYtNDllOS1iYWZmLTdjNDc5MDllMjY4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQwNjE5NDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNzVmYWE1M2ZjYWZkNTNlNTdmODhhN2RiYzA1ZmI2NzQ1YWNmZjI1OTYzNWMzOGNhN2Y1Mzk2MDkxNWE3ZTI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Y3KcX2ydNWWBzX4Kq_fFHN2Moms2FvPLLuy99ZWm30w">

没 get 到，什么方案？这个 Memo 还是这个笔记的布局？

JinliG committed:
perf: optimize doc reader

JinliG committed:
chore: .lock update

JinliG commented:
need review

Pleasurecruise commented:
conflicts麻烦处理一下

JinliG committed:
feat: remove auto input message

JinliG commented:
> conflicts麻烦处理一下

done

Pleasurecruise commented:
需要合并一下当前的主分支再看看能不能使用 因为一周前消息的类型进行了重构可能会存在问题

JinliG committed:
perf: remove console

JinliG committed:
Merge branch 'main' into feature/jinli/attach-document

Pleasurecruise commented:
整理一下我个人的一些建议：
1.加一个退出reader的button
2.页面放大的情况下可以拖动
3.具体关联页的选择似乎没有起作用？还是读取的整个文件

JinliG commented:
> 整理一下我个人的一些建议： 1.加一个退出reader的button 2.页面放大的情况下可以拖动 3.具体关联页的选择似乎没有起作用？还是读取的整个文件

感谢
之前的写法有点脏，这个 Reader 是我二次开发的副产物之一，还有些 bug，其中包括第三点，目前已经修复。
本周计划把新的代码 pick 过来，优化并兼容下 message chunk 的重构。然后再提 review。

JinliG committed:
feat: 优化阅读器交互&增加

JinliG committed:
fix: translate

JinliG committed:
fix: translate

JinliG committed:
Merge branch 'main' into feature/jinli/attach-document

JinliG committed:
fix: remove console

JinliG committed:
refactor: remove unused code

Pleasurecruise commented:
从交互上提几点改进的建议

在页面放大的情况下能进行拖动吗 比如下面这种情况右侧的部分就看不到了

<img src="https://github.com/user-attachments/assets/60045097-7d9e-4d8d-96a4-7ebf9f65834f" width="500" />

要去除inputbar的这个关联 只能从助手设置中删除关联的文档......吗

<img src="https://github.com/user-attachments/assets/48f22914-6b49-4fa9-b560-6dc3f1e1d9c0" width="500" />

还有 原先的左右切换按钮是去掉了......？

JinliG commented:
@Pleasurecruise 
阅读器右上角的参考资料 Popover 里，可以点击关闭阅读器。这里主要是因为 UI 上放操作按钮的空间不大够了，然后左右切换分栏的功能也是，放不下了。
![image](https://github.com/user-attachments/assets/67e71a2b-d8c3-468c-893a-f0569340d3ac)

JinliG commented on src/renderer/src/pages/home/PdfReader/index.tsx:
```diff
@@ -0,0 +1,330 @@
+
+  return (
+    <Container>
+      {pdfFile && (
```
期望包含哪些基础功能？

JinliG committed:
style: 调整操作栏样式

JinliG committed:
style: 调整 tip 位置防遮挡

DeJeune commented on src/renderer/src/i18n/locales/ja-jp.json:
```diff
@@ -2048,6 +2054,13 @@
         }
       }
+    },
+    "reader": {
```
翻译脚本没有维护简繁日英俄这五种语言
check:i18n只是识别哪些没被翻译
目前i18n翻译还没上CI，比较快捷的方式是cursor选中未翻译的部分(to be translated)，然后ctrl + K 让它翻译

JinliG commented on package.json:
```diff
@@ -94,6 +94,7 @@
     "officeparser": "^4.1.1",
     "os-proxy-config": "^1.1.2",
     "proxy-agent": "^6.5.0",
+    "react-pdf": "^9.2.1",
```
目前 lock 文件里没看到有其他依赖 pdfjs 的库，后面出现了再 resolve 吧

JinliG committed:
docs: 更新国际化

JinliG committed:
Merge branch 'main' into feature/jinli/attach-document

JinliG committed:
chore: resolve conflict problem

DeJeune commented on package.json:
```diff
@@ -84,7 +84,6 @@
     "@cherrystudio/embedjs-loader-web": "^0.1.31",
     "@cherrystudio/embedjs-loader-xml": "^0.1.31",
     "@cherrystudio/embedjs-ollama": "^0.1.31",
-    "@cherrystudio/embedjs-openai": "^0.1.31",
```
这里是不是删错了

JinliG committed:
chore: resolve conflict problem

JinliG committed:
build: move react-pdf to devDeps

JinliG committed:
docs: 调整国际化

JinliG committed:
refactor: 重构 Reader 提取多种阅读器

JinliG commented:
> 这个功能可以直接点击引用跳转到pdf的对应位置吗？

目前不行

DeJeune commented:
你好解决一下冲突

JinliG committed:
Merge branch 'main' into feature/jinli/attach-document

JinliG committed:
fix: update markdown reader

JinliG commented:
@alephpiece 
谢，react mk 问题已处理。渲染问题我注意到了，目测是 js 计算 pageWith 写法导致的，暂时还没时间去优化。。

JinliG committed:
Merge branch 'main' into feature/jinli/attach-document

DeJeune commented:
滚动时页码不跟着变化

JinliG committed:
feat: 页码监听

JinliG commented:
> <img alt="image" width="639" src="https://private-user-images.githubusercontent.com/67425183/460384333-325369ce-21be-49d1-a3a3-c57b3956c6a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE5NjI5OTEsIm5iZiI6MTc1MTk2MjY5MSwicGF0aCI6Ii82NzQyNTE4My80NjAzODQzMzMtMzI1MzY5Y2UtMjFiZS00OWQxLWEzYTMtYzU3YjM5NTZjNmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA4VDA4MTgxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhODZmZDg0YWJjYWIwNDM3MGM0ZmYxYmY5MjA0Njc1M2JjYmQ1OGFmN2E1MmViZmEyYmM0ZTJlMTkyY2Y5ZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RiMtCHKkNzZ14zwxZ6xTBccmmPlfsCtqgqNPf5PPrEA">
> 你好，xeLaTex编译的pdf，中文无法预览

react-pdf 组件缺少对 LaTeX  的支持，这个目测需要自定义渲染解决

JinliG committed:
Merge branch 'main' into feature/jinli/attach-document

eeee0717 committed:
fix: pdf-dist not work

DeJeune committed:
feat(file): read binary file

Pleasurecruise committed:
refactor: file selection and update upload icon

Moved the onSelectFile function inside the renderFileList callback and updated its dependencies for better encapsulation. Replaced the UploadOutlined icon with RefreshIcon for the reupload button. Also filtered files to only include documents when fetching from the database.

DeJeune committed:
fix: warning

DeJeune committed:
Merge branch 'feature/jinli/attach-document' of https://github.com/jinlig/cherry-studio into feature/jinli/attach-document

DeJeune committed:
fix: i18n

Pleasurecruise committed:
feat: add drag-to-pan functionality to PDF reader

Introduced state and event handlers to enable dragging the PDF document when zoomed in (scale > 1). Updated the DocumentReader component to reflect drag state with appropriate cursor changes and event bindings.

Pleasurecruise commented:
https://github.com/CherryHQ/cherry-studio/pull/5193/commits/c8eb7002f19134c6fed15d0eaee53bff37ddfe92 加了放大可拖动 可能需要看一下有啥没考虑到的

DeJeune committed:
feat: enhance PDF reader with text selection and drag mode toggle

Added new features to the PDF reader, including the ability to toggle between drag mode and text selection mode. Updated the user interface to reflect these changes and modified the internationalization files to support new strings for multiple languages.

0xfullex commented:
解决一下冲突

DeJeune committed:
Merge branch 'main' into feature/jinli/attach-document

DeJeune committed:
refactor: update import statement for CodeBlockView to named import

DeJeune committed:
feat(i18n): update Japanese, Russian, and Traditional Chinese translations for selection tool and reader features

Pleasurecruise committed:
Update index.tsx

beyondkmp commented on electron.vite.config.ts:
```diff
@@ -1,8 +1,17 @@
+const cMapsDir = normalizePath(path.join(path.dirname(require.resolve('pdfjs-dist/package.json')), 'cmaps'))
+const standardFontsDir = normalizePath(
+  path.join(path.dirname(require.resolve('pdfjs-dist/package.json')), 'standard_fonts')
+)
```
是直接不要打包进来，对吗

DeJeune committed:
Merge branch 'main' into feature/jinli/attach-document

DeJeune committed:
refactor: streamline topic management in Chat and AttachmentPreview components

- Simplified the Reader component by removing unnecessary props and integrating active topic management.
- Updated AttachmentPreview to utilize the active topic hook for better state handling.
- Refactored event handlers to improve clarity and performance in managing attached documents and pages.

DeJeune committed:
refactor: enhance topic management in AttachmentPreview and Reader components

- Updated AttachmentPreview to utilize the useTopic hook for improved state handling of topics.
- Refactored Reader component to replace activeTopic with _topic, ensuring consistent topic management.
- Streamlined event handlers for updating attached pages, enhancing clarity and performance.

DeJeune committed:
Merge branch 'main' into feature/jinli/attach-document

DeJeune committed:
Merge branch 'main' into feature/jinli/attach-document

DeJeune committed:
feat(Inputbar): integrate useTopic hook for topic management

- Added the useTopic hook to manage topic state within the Inputbar component.
- Updated references to topic.attachedPages to use the new _topic variable for improved clarity and consistency.
- Adjusted effect dependencies to include _topic for accurate state updates.

EurFelux commented:
1. 既然现在只支持pdf的话，能不能在从文件中选择的时候也筛一下扩展名？参考资料中也是，选择其他格式的文件也会直接发送出去。像是图片之类的，原本inputbar有按模型做上传图片限制，这里都可以直接突破模型限制发送出去了。
2. 页面选择功能开启后，每一页右上角会出现一个框，但不是很显眼。我一开始还很困惑这个功能怎么用。看能不能给这个checkbox出现的时候加一点视觉动效，引导一下用户？
3. “正在关联”的意思是“已关联”对吗？文案用进行时容易让人误解关联过程一直没完成。我等半天还以为应该有个进度条什么的。
4. 缩放后会导致一段时间内文档全部变成空白，等上一会儿才能重新渲染。
5. 缩小后会得到很微妙的数字，哎浮点数
6. 缩小后文档左对齐，右边留白的，看起来很奇怪
<img width="701" height="729" alt="image" src="https://github.com/user-attachments/assets/4095423b-e5db-4a4a-a840-455eeb82fbfe" />
7. 放得更小就更奇怪，整个宽度是动不了的，反而高度一直在降
<img width="692" height="726" alt="image" src="https://github.com/user-attachments/assets/113cb287-1599-4fd6-ab0e-bf0785b7a772" />
8. 拖拽模式在缩小的时候显得没有什么意义？只是变得能够通过拖拽来滚动而已。如果只是这样的功能，那在100%的时候不允许拖拽的设计也变得奇怪了。
9. 开启拖拽模式后，光标移动到文本上时仍然会变成文本选择模式，导致无法拖拽。
10. 一般pdf reader这个地方应该是随页面滚动变化的，这样用户能够知道现在是pdf的第几页了
<img width="145" height="79" alt="image" src="https://github.com/user-attachments/assets/444eae66-54d5-4205-b6fc-a88f1707069a" />
11. 用户无法知道当前浏览的页数的同时，如果前面选择了特定页数，只能在输入栏中看到选择了第几页，但无法从阅读器中直观地看到是哪一页。想要知道选择了哪些页，就必须重新打开页面选择，然后观察哪些页右上角选中了，这个交互体验很糟糕。
12. 选择页数后，若不启用关联，选择页数就不会发送出去。这个应该设计的时候就是这样是吧？但是跟之前讨论里说过的那样，现在选择页数没有作用，实际上还是把整个文档发出去了。
13. 开启/关闭关联会导致文档重新渲染
14. 新建话题会导致已选择的页数清空。选择的页数是话题的状态，而不是输入栏的吗？但这个状态显示在输入栏，像是文件上传和提及模型这样的状态都是不会随话题切换而变动的，这里设计上存在不一致。

Pleasurecruise committed:
Merge branch 'main' into feature/jinli/attach-document

Pleasurecruise committed:
fix: merge error

kangfenmao commented:
Any update?
